diff -ruN radvd-0.7.2.orig/scanner.l radvd-0.7.2/scanner.l
--- radvd-0.7.2.orig/scanner.l	2002-07-11 23:26:16.000000000 +0200
+++ radvd-0.7.2/scanner.l	2004-02-22 16:53:10.000000000 +0100
@@ -34,7 +34,7 @@
 addr3		({hexdigit}{1,4}":"){7}{hexdigit}{1,4}
 addr		({addr1}|{addr2}|{addr3}|"::")
 whitespace	([ \t])+
-string		[a-z]([a-z]|{digit}|_)*([:.]{digit}+)?
+string		[a-z]([a-z:._-]|{digit})*([:.]{digit}+)?
 %%
 
 #.*$			{/* ignore comments */}
